server/src/main/java/org/opensearch/common/blobstore/transfer/RemoteTransferContainer.java

```
@@ -0,0 +1,252 @@
+        return ((partNo, size, position) -> {
+            assert inputStreams.get() != null : "expected inputStreams to be initialised";
+            return getMultipartStreamSupplier(partNo, size, position).get();
+        });
+    }
+
+    /**
+     * OffsetRangeInputStreamSupplier is used to get the offset based input streams at runtime
+     */
+    public interface OffsetRangeInputStreamSupplier {
+        OffsetRangeInputStream get(long size, long position) throws IOException;
+    }
+
+    interface LocalStreamSupplier<Stream> {
+        Stream get() throws IOException;
+    }
+
+    private LocalStreamSupplier<InputStreamContainer> getMultipartStreamSupplier(
+        final int streamIdx,
+        final long size,
+        final long position
+    ) {
+        return () -> {
+            try {
+                OffsetRangeInputStream offsetRangeInputStream = offsetRangeInputStreamSupplier.get(size, position);
+                InputStream inputStream = !isRemoteDataIntegrityCheckPossible()
+                    ? new ResettableCheckedInputStream(offsetRangeInputStream, fileName)
+                    : offsetRangeInputStream;
+                Objects.requireNonNull(inputStreams.get())[streamIdx] = inputStream;
+
+                return new InputStreamContainer(inputStream, size);
+            } catch (IOException e) {
+                log.error("Failed to create input stream", e);
+                throw e;
+            }
+        };
+    }
+
+    private boolean isRemoteDataIntegrityCheckPossible() {
+        return isRemoteDataIntegritySupported && !areInputStreamsDecorated;
+    }
+
+    private void finalizeUpload(boolean uploadSuccessful) {
+        if (isRemoteDataIntegrityCheckPossible()) {
+            return;
+        }
+
+        if (uploadSuccessful) {
+            long actualChecksum = getActualChecksum();
+            if (actualChecksum != expectedChecksum) {
+                throw new RuntimeException(
```

Why we need to wrap the exception here? The finalizeUpload should be able to throw IOException

---

The reason we structured it like this is that this method is called from inside a [lambda in the plugin](https://github.com/opensearch-project/OpenSearch/pull/7217/files#diff-a472eda8bf5f051224172440b745603502d3545f8326b28d6014a79d5a833cd5R166). We can either throw a runtime exception from here, or we'll need to use something like a `CheckedBiConsumer` here and handle it inside the lambda. What do you think is the better way of handling this?

---

That's should be the caller concern, the code should not adapt for every caller's need, please make IOException explicit in signature and throw the unwrapped one, thank you.